// @expo/build-tools/dist/platforms/android/generic/builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const project_1 = require("../../../utils/project");
const spawn_1 = __importDefault(require("../../../utils/spawn"));
const buildConfig_1 = require("../utils/buildConfig");
async function buildGeneric(ctx) {
    const { job, logger } = ctx;
    const { buildDirectory, nativeProjectDirectory } = await project_1.setup(ctx);
    logger.info('Writing keystore to the project\'s directory');
    const { credentials: { keystore } } = job;
    const keystorePath = path_1.default.join(buildDirectory, keystore.keystorePath);
    await buildConfig_1.saveKeystore(keystore, keystorePath);
    logger.info('Starting gradle build');
    const keystoreEnvs = {
        KEYSTORE_PASSWORD: keystore.keystorePassword,
        KEY_ALIAS: keystore.keyAlias,
        KEY_PASSWORD: keystore.keyPassword,
    };
    logger.debug(`Running './gradlew ${job.gradleCommand}' in ${nativeProjectDirectory}`);
    await spawn_1.default('./gradlew', [job.gradleCommand], {
        cwd: nativeProjectDirectory,
        logger,
        env: {
            ...process.env,
            ...keystoreEnvs,
        },
    });
    return path_1.default.join(buildDirectory, job.artifactPath);
}
exports.default = buildGeneric;
