// @expo/build-tools/dist/buildConfig/converter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_extra_1 = __importDefault(require("fs-extra"));
async function prepareAndroidJobCredentials(credentials) {
    const keystoreDataBase64 = await fs_extra_1.default.readFile(credentials.android.keystore.keystorePath, 'base64');
    const keystore = {
        ...credentials.android.keystore,
        keystoreDataBase64,
    };
    const androidJobCredentials = {
        keystore,
    };
    return androidJobCredentials;
}
exports.prepareAndroidJobCredentials = prepareAndroidJobCredentials;
async function prepareiOSJobCredentials(credentials) {
    const provisioningProfileData = await fs_extra_1.default.readFile(credentials.ios.provisioningProfilePath, 'base64');
    const distributionCertificateData = await fs_extra_1.default.readFile(credentials.ios.distributionCertificate.path, 'base64');
    const iOSJobCredentials = {
        provisioningProfileBase64: provisioningProfileData,
        distributionCertificate: {
            dataBase64: distributionCertificateData,
            password: credentials.ios.distributionCertificate.password,
        },
    };
    return iOSJobCredentials;
}
exports.prepareiOSJobCredentials = prepareiOSJobCredentials;
