// @expo/build-tools/dist/buildConfig/schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("@hapi/joi"));
const AndroidBuildConfigSchema = joi_1.default.object().keys({
    keystore: joi_1.default.object().keys({
        keystorePath: joi_1.default.string().required(),
        keystorePassword: joi_1.default.string().required(),
        keyAlias: joi_1.default.string().required(),
        keyPassword: joi_1.default.string().required(),
    }).required(),
});
const iOSBuildConfigSchema = joi_1.default.object().keys({
    provisioningProfilePath: joi_1.default.string().required(),
    distributionCertificate: joi_1.default.object().keys({
        path: joi_1.default.string().required(),
        password: joi_1.default.string().required(),
    }).required(),
});
const BuildConfigSchema = joi_1.default.object().keys({
    android: AndroidBuildConfigSchema,
    ios: iOSBuildConfigSchema,
});
exports.default = BuildConfigSchema;
