// @expo/build-tools/src/buildConfig/converter.ts
import fs from 'fs-extra';

import * as Android from '../platforms/android';
import * as iOS from '../platforms/ios';
import { BuildConfig } from './types';

async function prepareAndroidJobCredentials(credentials: BuildConfig): Promise<Android.Credentials> {
  const keystoreDataBase64 = await fs.readFile(credentials.android.keystore.keystorePath, 'base64');
  const keystore = {
    ...credentials.android.keystore,
    keystoreDataBase64,
  };
  const androidJobCredentials = {
    keystore,
  };
  return androidJobCredentials;
}

async function prepareiOSJobCredentials(credentials: BuildConfig): Promise<iOS.Credentials> {
  const provisioningProfileData = await fs.readFile(credentials.ios.provisioningProfilePath, 'base64');
  const distributionCertificateData = await fs.readFile(credentials.ios.distributionCertificate.path, 'base64');
  const iOSJobCredentials = {
    provisioningProfileBase64: provisioningProfileData,
    distributionCertificate: {
      dataBase64: distributionCertificateData,
      password: credentials.ios.distributionCertificate.password,
    },
  };
  return iOSJobCredentials;
}

export { prepareAndroidJobCredentials, prepareiOSJobCredentials };
