// @expo/build-tools/src/platforms/ios/utils/gymfile.ts
import path from 'path';

import envsub from 'envsub';
import fs from 'fs-extra';

import { Credentials } from '../credentials';

const TEMPLATE_FILE_PATH = path.join(__dirname, '../../../../templates/Gymfile.template');

interface GymfileOptions {
  nativeProjectDirectory: string;
  outputName: string;
  outputDirectory: string;
}

async function createIfNotExists(
  credentials: Credentials,
  { nativeProjectDirectory, outputName, outputDirectory }: GymfileOptions,
) {
  const outputFile = path.join(nativeProjectDirectory, 'Gymfile');

  if (await fs.pathExists(outputFile)) {
    return false;
  } else {
    await envsub({
      templateFile: TEMPLATE_FILE_PATH,
      outputFile,
      options: {
        envs: [
          { name: 'BUNDLE_IDENTIFIER', value: credentials.bundleIdentifier },
          { name: 'PROVISIONING_PROFILE_UUID', value: credentials.provisioningProfileUUID },
          { name: 'KEYCHAIN_PATH', value: credentials.keychainPath },
          { name: 'OUTPUT_DIRECTORY', value: outputDirectory },
          { name: 'OUTPUT_NAME', value: outputName },
        ],
        syntax: 'handlebars',
      },
    });
    return true;
  }
}

export { createIfNotExists };
