// @expo/build-tools/dist/platforms/android/baseJob.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("@hapi/joi"));
const constants_1 = require("../../constants");
const AndroidCredentialsSchema = joi_1.default.object().keys({
    keystore: joi_1.default.object().keys({
        keystoreDataBase64: joi_1.default.string().required(),
        keystorePath: joi_1.default.string().required(),
        keystorePassword: joi_1.default.string().required(),
        keyAlias: joi_1.default.string().required(),
        keyPassword: joi_1.default.string().required(),
    }).required(),
});
const BaseAndroidJobSchema = joi_1.default.object().keys({
    platform: joi_1.default.string().valid(constants_1.Platform.Android).required(),
    projectUrl: joi_1.default.string().uri().required(),
    nativeProjectDirectory: joi_1.default.string().default('android'),
    credentials: AndroidCredentialsSchema.required(),
});
exports.BaseAndroidJobSchema = BaseAndroidJobSchema;
