// @expo/build-tools/src/platforms/android/baseJob.ts
import Joi from '@hapi/joi';

import { Platform } from '../../constants';

interface BaseAndroidJob {
  platform: Platform.Android;
  projectUrl: string;
  nativeProjectDirectory: string;
  credentials: AndroidCredentials;
}

interface AndroidCredentials {
  keystore: AndroidKeystore;
}

const AndroidCredentialsSchema = Joi.object().keys({
  keystore: Joi.object().keys({
    keystoreDataBase64: Joi.string().required(),
    keystorePath: Joi.string().required(),
    keystorePassword: Joi.string().required(),
    keyAlias: Joi.string().required(),
    keyPassword: Joi.string().required(),
  }).required(),
});

interface AndroidKeystore {
  keystoreDataBase64: string;
  keystorePath: string;
  keystorePassword: string;
  keyAlias: string;
  keyPassword: string;
}

const BaseAndroidJobSchema = Joi.object().keys({
  platform: Joi.string().valid(Platform.Android).required(),
  projectUrl: Joi.string().uri().required(),
  nativeProjectDirectory: Joi.string().default('android'),
  credentials: AndroidCredentialsSchema.required(),
});

export { BaseAndroidJob, BaseAndroidJobSchema, AndroidCredentials, AndroidKeystore };
