// expo-cli/build/credentials/credentials.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EXPO_WILL_GENERATE = exports.keystoreSchema = exports.pushKeySchema = exports.distCertSchema = void 0;
// 
// iOS
//
const distCertSchema = {
  id: 'distributionCert',
  canReuse: true,
  name: 'Apple Distribution Certificate',
  required: ['certP12', 'certPassword'],
  questions: {
    certP12: {
      question: 'Path to P12 file:',
      type: 'file',
      base64Encode: true
    },
    certPassword: {
      type: 'password',
      question: 'P12 password:'
    }
  }
};
exports.distCertSchema = distCertSchema;
const pushKeySchema = {
  id: 'pushKey',
  canReuse: true,
  name: 'Apple Push Notifications service key',
  required: ['apnsKeyP8', 'apnsKeyId'],
  questions: {
    apnsKeyP8: {
      type: 'file',
      question: 'Path to P8 file:'
    },
    apnsKeyId: {
      type: 'string',
      question: 'Key ID:'
    }
  }
};
exports.pushKeySchema = pushKeySchema;
const provisioningProfileSchema = {
  id: 'provisioningProfile',
  name: 'Apple Provisioning Profile',
  required: ['provisioningProfile'],
  dependsOn: 'distributionCert',
  questions: {
    provisioningProfile: {
      type: 'file',
      question: 'Path to .mobile provisioning profile:',
      base64Encode: true
    }
  }
};
const pushCertSchema = {
  id: 'pushCert',
  name: 'Apple Push Notifications certificate',
  required: ['pushP12', 'pushPassword'],
  deprecated: true,
  migrationDocs: 'https://docs.expo.io/versions/latest/distribution/building-standalone-apps/#switch-to-push-notification-key-on-ios'
}; //
// Android
//

const keystoreSchema = {
  id: 'keystore',
  name: 'Android Keystore',
  required: ['keystore', 'keystorePassword', 'keyAlias', 'keyPassword'],
  questions: {
    keystore: {
      question: 'Path to the Keystore file.',
      type: 'file',
      base64Encode: true
    },
    keystorePassword: {
      question: 'Keystore password',
      type: 'password'
    },
    keyAlias: {
      question: 'Key alias',
      type: 'string'
    },
    keyPassword: {
      question: 'Key password',
      type: 'password'
    }
  }
};
exports.keystoreSchema = keystoreSchema;
const EXPO_WILL_GENERATE = 'EXPO_PLEASE_GENERATE_THIS_FOR_ME';
exports.EXPO_WILL_GENERATE = EXPO_WILL_GENERATE;
