// @expo/webpack-config/webpack/loaders/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var createAllLoaders_1 = require("./createAllLoaders");
exports.imageLoaderRule = createAllLoaders_1.imageLoaderRule;
exports.fallbackLoaderRule = createAllLoaders_1.fallbackLoaderRule;
exports.styleLoaderRule = createAllLoaders_1.styleLoaderRule;
exports.getBabelLoaderRuleFromEnv = createAllLoaders_1.getBabelLoaderRuleFromEnv;
exports.getBabelLoaderRule = createAllLoaders_1.getBabelLoaderRule;
exports.getHtmlLoaderRule = createAllLoaders_1.getHtmlLoaderRule;
exports.getAllLoaderRules = createAllLoaders_1.getAllLoaderRules;
exports.createAllLoaders = createAllLoaders_1.default;
var createBabelLoader_1 = require("./createBabelLoader");
exports.createBabelLoader = createBabelLoader_1.default;
__export(require("./createBabelLoader"));
var createFontLoader_1 = require("./createFontLoader");
exports.createFontLoader = createFontLoader_1.default;
